### PR TITLE
 Add optional storages

### PR DESCRIPTION
**Pull request by jarbitlira**
I add this feature because if I use cordova/phonegab, cookies aren't detected and this plugin doesn't work correctly. For this reason I add the optional storage to localStorage and sessionStorage

### CODE REVIEW

File: src/providers/oauth-provider.js [hunk 1]
```diff
@@ -167,8 +167,8 @@ function OAuthProvider() {
         };
 
         return $http.post(`${config.baseUrl}${config.grantPath}`, data, options).then((response) => {
+
```
Comment (ruipenso) [inline src/providers/oauth-provider.js]: Revert this changes, please.

File: src/providers/oauth-token-provider.js [hunk 2]
```diff
@@ -10,8 +10,10 @@ import angular from 'angular';
  */
 
 function OAuthTokenProvider() {
+  var storage;
   var config = {
     name: 'token',
+    storage: 'cookies',//cookies,localStorage,sessionStorage
```
Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: Add a space before and after the comment `//` and between commas.

File: src/providers/oauth-token-provider.js [hunk 3]
```diff
@@ -49,7 +51,7 @@ function OAuthTokenProvider() {
        */
 
       set token(data) {
-        return ipCookie(config.name, data, config.options);
+        return setToken(data);
```
Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: Can we move this to a separate service?

File: src/providers/oauth-token-provider.js [hunk 4]
```diff
@@ -101,10 +103,71 @@ function OAuthTokenProvider() {
...
+
     }
 
+    /**
+     * setToken
+     *
+     * @param data
+     * @returns {*}
+     */
+    var setToken = function(data) {
```
Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: Add a new line after the comment block and use ES6, e.g.,

```
setToken(data) {
}
```

Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: BTW, I think it would be better to move this to a different service.

File: src/providers/oauth-token-provider.js [hunk 5]
```diff
@@ -101,10 +103,71 @@ function OAuthTokenProvider() {
...
+     storage = config.storage.toLowerCase();
+      switch (storage) {
+       case 'cookies':
+        return ipCookie(config.name, data, config.options);
+       case 'localstorage':
+        return localStorage.setItem(config.name,JSON.stringify(data));
```
Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: Use `$window.localStorage`.

(Add a space after comma)

File: src/providers/oauth-token-provider.js [hunk 6]
```diff
@@ -101,10 +103,71 @@ function OAuthTokenProvider() {
...
+      }
+    };
+
+    /**
+     * getToken
+     *
+     * @returns {*}
+     */
+    var getToken = function() {
+     storage = config.storage.toLowerCase();
```
Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: Wrong indentation.

File: src/providers/oauth-token-provider.js [hunk 7]
```diff
@@ -101,10 +103,71 @@ function OAuthTokenProvider() {
...
+      switch (storage) {
+       case 'cookies':
+        return ipCookie(config.name);
+       case 'localstorage':
+        return JSON.parse(localStorage.getItem(config.name));
+       case 'sessionstorage':
+        return JSON.parse(sessionStorage.getItem(config.name));
+       default :
```
Comment (ruipenso) [inline src/providers/oauth-token-provider.js]: Remove extra space.

**Comment by ruipenso** [comment]
Hi @jarbitlira!

Looks great.

It would be great if you could create a separate service to handle `setToken`, `getToken` and `removeToken` to get a better abstraction. WDYT?

Also, I added some comments... Mostly, code standards - try to maintain the code standards already in place.

Thanks  :)

**Comment by joaogranado** [comment]
Hi @jarbitlira could you add tests please?

File: src/providers/oauth-token-provider.js (hunk 5 above)
Comment (joaogranado) [inline src/providers/oauth-token-provider.js]: I would suggest to use `angular.toJson()` method instead of `JSON.stringify`.

File: src/providers/oauth-token-provider.js [hunk 8]
```diff
@@ -101,10 +103,71 @@ function OAuthTokenProvider() {
...
+     *
+     * @returns {*}
+     */
+    var getToken = function() {
+     storage = config.storage.toLowerCase();
+      switch (storage) {
+       case 'cookies':
+        return ipCookie(config.name);
+       case 'localstorage':
+        return JSON.parse(localStorage.getItem(config.name));
```
Comment (joaogranado) [inline src/providers/oauth-token-provider.js]: Use `angular.fromJson` instead of `JSON.parse`.

File: src/providers/oauth-token-provider.js (hunk 4 above)
Comment (joaogranado) [inline src/providers/oauth-token-provider.js]: I agree with @ruipenso, something like `oauthStorageService`, which would abstract the storage of the token.

**Comment by joaogranado** [comment]
Please rename the PR from **Feature/optional storages** to  **Add optional storages**.

**Comment by jarbitlira** [comment]
Hi everyone, thanks for yours comments, it's my first time with ES6. 

can you give me a guideline about code standards that are you using?

I go to enhance the code, but first could you tell me, what do you think about this feature?

**Comment by ruipenso** [comment]
@jarbitlira The feature is good.

We should consider to switch [ivpusic/angular-cookie](https://github.com/ivpusic/angular-cookie) by [grevory/angular-local-storage](https://github.com/grevory/angular-local-storage) which support multiple storage types, but still misses [cookie secure flag option](https://github.com/grevory/angular-local-storage/pull/185).

But, meanwhile, we can add this functionality with your PR.

Regarding the code standards... We don't have them written anywhere. So, you'll need to take a look at the code and try to figure up the best you can. Sorry about that.

**Comment by ugurerkan** [comment]
Is there any planned date for merge this PR?

**Comment by smarts** [comment]
I think it hasn't been merged yet because no tests have been added.

**Comment by adozenlines** [comment]
+1 right now the auth response isn't being serialized to the $cookie store ...

**Comment by jonkoops** [comment]
I've gone ahead and made a [separate pull request](https://github.com/oauthjs/angular-oauth2/pull/95) that replaces ngCookies with sessionStorage. sessionStorage works basically everywhere so we don't need to fall back to cookies. As for choosing between session and local storage we can tackle at a later point in time.

**Comment by joaogranado** [comment]
Closing this PR based on what was said at #95 and #97.
